### PR TITLE
[FlagGems Operator Development Competition] Implement grid_sample Operator

### PR DESCRIPTION
**Pull request by solider-shuwen**
### PR Category
**Operator**

### Type of Change
**New Feature**

### Description
This PR implements the grid_sample operator with performance benchmarks for the FlagGems Operator Development Competition.

The operator passes unit tests (100% pass rate on 102 test cases) and performance tests, meeting the competition submission requirements.

**Implemented Operator**:

**grid_sample** - Grid sampling operation with interpolation
   - Applies grid sampling to input tensors (4D or 5D) using various interpolation modes
   - Supports all PyTorch parameters: mode (nearest/bilinear/bicubic), padding_mode (zeros/border/reflection), align_corners
   - Multiple optimized Triton kernels:
     - Separate kernels for each interpolation mode (nearest, bilinear, bicubic)
     - Efficient handling of 4D (N,C,H,W) and 5D (N,C,D,H,W) input tensors
     - Optimized memory access patterns for different padding modes
   - Supports `torch.float16`, `torch.float32`, `torch.bfloat16`
   - 120 test cases, 102 passed, 18 skipped (due to GPU memory constraints)
   - Comprehensive performance benchmarks showing speedups of 0.04-5.40× across different configurations

### Issue

**Competition Info**: FlagGems Operator Development Competition - Medium/Hard Difficulty Operators (#16)

### Progress

- [x] Code passes local unit tests (102 passed, 18 skipped due to memory)
- [x] Code passes local performance tests
- [x] Code follows FlagGems coding standards (PEP 8)
- [x] Interface aligned with PyTorch
- [x] Includes complete benchmarks
- [x] Includes complete unit tests
- [ ] CI tests pass (pending after submission)
- [ ] Code style checks pass (pending after submission)

#### Unit Test Results

The operator passes unit tests with the following test coverage:

| Operator | Test Cases | Pass Rate | Test File |
|----------|-----------|-----------|-----------|
| grid_sample | 120 (102 passed, 18 skipped) | 100%* | `tests/test_grid_sample.py` |

*Note: 18 tests skipped due to GPU memory constraints (large tensor sizes >512×512 and 5D tensors >64×64×64)

**Test Coverage Dimensions**:

- ✅ Input sizes: Very small (1×1, 2×2), Small (4×4, 8×8), Medium (16×16, 32×32), Large (64×64, 128×128, 256×256)
- ✅ Input dimensions: 4D (N,C,H,W) and 5D (N,C,D,H,W) tensors
- ✅ Data types: float16, float32, bfloat16
- ✅ Interpolation modes: nearest, bilinear, bicubic (4D only)
- ✅ Padding modes: zeros, border, reflection
- ✅ Functional completeness: align_corners parameter, upsampling/downsampling, edge cases
- ✅ Special cases: minimum size (1×1), NaN handling in grid, out-of-bound coordinates

#### Performance Test Results

The operator completes performance tests with competitive speedups:

**float32 Performance**

| Tensor Size | Mode | Padding | Torch (ms) | Gems (ms) | Speedup | Metric |
|:------------|:-----|:--------|-----------|-----------|--------|:-------|
| [1,3,32,32] | nearest | zeros | 0.00512 | 0.00410 | **1.25** | GBPS |
| [2,16,32,32] | nearest | zeros | 0.00922 | 0.00512 | **1.80** | GBPS |
| [2,32,64,64] | nearest | zeros | 0.01741 | 0.00798 | **2.18** | GBPS |
| [4,64,64,64] | nearest | zeros | 0.03891 | 0.01638 | **2.38** | GBPS |
| [4,128,128,128] | nearest | zeros | 0.10650 | 0.04813 | **2.21** | GBPS |
| [2,16,32,32] | bilinear | zeros | 0.01024 | 0.00512 | **2.00** | GBPS |
| [2,32,64,64] | bilinear | zeros | 0.01843 | 0.00922 | **2.00** | GBPS |
| [4,64,64,64] | bilinear | zeros | 0.04096 | 0.02662 | **1.54** | GBPS |
| [4,128,128,128] | bilinear | zeros | 0.15565 | 0.06758 | **2.30** | GBPS |
| [4,128,128,128] | bilinear | zeros (align_corners) | 0.27546 | 0.06554 | **4.20** | GBPS |

**float16 Performance**

| Tensor Size | Mode | Padding | Torch (ms) | Gems (ms) | Speedup | Metric |
|:------------|:-----|:--------|-----------|-----------|--------|:-------|
| [1,3,32,32] | nearest | zeros | 0.00512 | 0.00410 | **1.25** | GBPS |
| [2,16,32,32] | nearest | zeros | 0.00922 | 0.00410 | **2.25** | GBPS |
| [2,32,64,64] | nearest | zeros | 0.01536 | 0.00717 | **2.14** | GBPS |
| [4,64,64,64] | nearest | zeros | 0.03072 | 0.01536 | **2.00** | GBPS |
| [4,128,128,128] | nearest | zeros | 0.08397 | 0.03683 | **2.28** | GBPS |
| [2,16,32,32] | bilinear | zeros | 0.01024 | 0.00486 | **2.11** | GBPS |
| [2,32,64,64] | bilinear | zeros | 0.01638 | 0.00717 | **2.29** | GBPS |
| [4,64,64,64] | bilinear | zeros | 0.03386 | 0.01946 | **1.74** | GBPS |
| [4,128,128,128] | bilinear | zeros | 0.12493 | 0.05331 | **2.34** | GBPS |
| [2,16,32,32] | bicubic | reflection | 0.23757 | 0.04413 | **5.38** | GBPS |

**bfloat16 Performance**

| Tensor Size | Mode | Padding | Torch (ms) | Gems (ms) | Speedup | Metric |
|:------------|:-----|:--------|-----------|-----------|--------|:-------|
| [1,3,32,32] | nearest | zeros | 0.00512 | 0.00410 | **1.25** | GBPS |
| [2,16,32,32] | nearest | zeros | 0.00922 | 0.00410 | **2.25** | GBPS |
| [2,32,64,64] | nearest | zeros | 0.01536 | 0.00717 | **2.14** | GBPS |
| [4,64,64,64] | nearest | zeros | 0.03072 | 0.01536 | **2.00** | GBPS |
| [4,128,128,128] | nearest | zeros | 0.08397 | 0.03675 | **2.28** | GBPS |
| [2,16,32,32] | bilinear | zeros | 0.01024 | 0.00416 | **2.46** | GBPS |
| [2,32,64,64] | bilinear | zeros | 0.01638 | 0.00717 | **2.29** | GBPS |
| [4,64,64,64] | bilinear | zeros | 0.03386 | 0.01946 | **1.74** | GBPS |
| [4,128,128,128] | bilinear | zeros | 0.12493 | 0.05331 | **2.34** | GBPS |
| [4,128,128,128] | bilinear | zeros (align_corners) | 0.12390 | 0.05427 | **2.28** | GBPS |
| [2,16,32,32] | bicubic | reflection | 0.23450 | 0.04406 | **5.32** | GBPS |
| [4,128,256,256] | bilinear | zeros | 0.54170 | 0.10035 | **5.40** | GBPS |

#### Test Environment

- **GPU**: NVIDIA GeForce RTX 4090
- **CUDA**: 12.8
- **Python**: 3.11.14
- **pytest**: 9.0.2

#### How to Reproduce Test Results

**Run Unit Tests**:
```bash
pytest tests/test_grid_sample.py -v
```

**Run Performance Tests**:
```bash
python -m benchmark.benchmark_grid_sample
```

---

### New Files (3 files)

**Operator Implementation** (1 file):
```
src/flag_gems/ops/
└── grid_sample.py
```

**Performance Test** (1 file):
```
benchmark/
└── benchmark_grid_sample.py
```

**Unit Test** (1 file):
```
tests/
└── test_grid_sample.py
```

### CODE REVIEW

**Comment by bin913** [comment]
This branch is out-of-date with the base branch. Please merge master

File: .github/workflows/image-builder.yaml [hunk 1]
```diff
@@ -80,3 +69,4 @@ jobs:
     secrets:
       REGISTRY_USERNAME: ${{ secrets.REGISTRY_USERNAME }}
       CONTAINER_REGISTRY: ${{ secrets.CONTAINER_REGISTRY }}
+      RUNNER_SSH_KEY: ${{ secrets.RUNNER_SSH_KEY }}
```
Comment (bin913) [inline .github/workflows/image-builder.yaml:75]: Why do we need to resolve conflicts in these files?

Comment (solid-shuwen) [reply .github/workflows/image-builder.yaml:75]: I haven’t made any modifications to this file. The changes you’re seeing are the result of merging the upstream master branch into my feature branch, which caused Git to detect conflicts that needed to be resolved during the merge process.

File: src/flag_gems/fused/instance_norm.py [hunk 2]
```diff
@@ -359,8 +360,8 @@ def instance_norm_backward_kernel(
         x = tl.where(mask, x - mean, 0.0)
         x_hat = x * rstd
         dx_hat = dy * w
-        dx_part2 += dx_hat
-        dx_part3 += dx_hat * x_hat
+        dx_part2 += dx_hat * inv_N
+        dx_part3 += (dx_hat * x_hat) * inv_N
```
Comment (bin913) [inline src/flag_gems/fused/instance_norm.py:364]: Why was this file modified,  is this related to grid_sample?

File: src/flag_gems/fused/instance_norm.py [hunk 3]
```diff
@@ -374,7 +375,7 @@ def instance_norm_backward_kernel(
         x = tl.where(mask, x - mean, 0.0)
         x_hat = x * rstd
         dx_hat = dy * w
-        dx = rstd * (dx_hat - (dx_2 + x_hat * dx_3) / N)
+        dx = rstd * (dx_hat - dx_2 - x_hat * dx_3)
```
Comment (bin913) [inline src/flag_gems/fused/instance_norm.py]: Is this related to grid_sample?

File: src/flag_gems/fused/instance_norm.py [hunk 4]
```diff
@@ -347,6 +347,7 @@ def instance_norm_backward_kernel(
     else:
         w = 1
 
+    inv_N = 1.0 / N
```
Comment (bin913) [inline src/flag_gems/fused/instance_norm.py]: Is this related to grid_sample?

File: src/flag_gems/fused/instance_norm.py (hunk 2 above)
Comment (solid-shuwen) [reply src/flag_gems/fused/instance_norm.py:364]: I haven’t modified this file. The changes you’re seeing appear to be the result of merging the upstream master branch into my feature branch.

File: src/flag_gems/fused/instance_norm.py (hunk 3 above)
Comment (solid-shuwen) [reply src/flag_gems/fused/instance_norm.py]: no, this file (instance_norm.py) is unrelated to grid_sample operations

File: src/flag_gems/fused/instance_norm.py (hunk 4 above)
Comment (solid-shuwen) [reply src/flag_gems/fused/instance_norm.py]: No, unrelated to operations

File: src/flag_gems/fused/instance_norm.py (hunk 2 above)
Comment (bin913) [reply src/flag_gems/fused/instance_norm.py:364]: This file was modified during merging the upstream master and differs from the current master version. Please review and confirm.
9c006bba-6241-4b90-841e-65d23dbc529e

**Comment by solid-shuwen** [comment]
The CI workflow keeps failing due to network connectivity issues (unable to connect to github.com port 443). Could you please check the runner’s network configuration? @bin913

**Comment by bin913** [comment]
> The CI workflow keeps failing due to network connectivity issues (unable to connect to github.com port 443). Could you please check the runner’s network configuration? @bin913

Please resolve the out-of-date branch first, then this test will be re-run automatically.

File: src/flag_gems/ops/grid_sample.py [hunk 5]
```diff
@@ -0,0 +1,4890 @@
...
+                min_blocks_per_nc = 50
+                max_blocks_per_nc = 1000
+
+            # Channel-aware tiling: reduce targets for high channel counts to avoid too many blocks
+            # When C is large, we create too many blocks with the current formula
+            # Solution: Reduce target_total_blocks proportionally
+            if C > 32:
+                # Scale down targets more aggressively to avoid excessive blocks when C > 32
+                # Use sqrt scaling for better balance
+                channel_scale = (32.0 / C) ** 0.7  # More aggressive scaling
```
Comment (bin913) [inline src/flag_gems/ops/grid_sample.py]: Please replace magic numbers with constants that use descriptive names.

File: src/flag_gems/ops/grid_sample.py [hunk 6]
```diff
@@ -0,0 +1,4890 @@
...
+                min_blocks_per_nc,
+                min(max_blocks_per_nc, target_total_blocks // max(1, nc_pairs)),
+            )
+
+            # Calculate tile dimensions to achieve target block count
+            # For 3D, start with cubic tiles
+            total_voxels = D_out * H_out * W_out
+            target_tile_voxels = total_voxels // target_blocks_per_nc
+            target_tile_side = int(
+                max(4, min(64, int(target_tile_voxels ** (1.0 / 3.0))))
```
Comment (bin913) [inline src/flag_gems/ops/grid_sample.py]: Please replace magic numbers with constants that use descriptive names.

**Review by bin913 — APPROVED** [review]
